pandas/_libs/lib.pyx

```diff
@@ -1089,6 +1100,79 @@ def is_float(obj: object) -> bool:
     return util.is_float_object(obj)
 
 
+cpdef bint is_hashable(object obj):
+    """
+    Return True if hash(obj) will succeed, False otherwise.
+
+    Some types will pass a test against collections.abc.Hashable but fail when
+    they are actually hashed with hash().
+
+    Distinguish between these and other types by trying the call to hash() and
+    seeing if they raise TypeError.
+
+    Returns
+    -------
+    bool
+
+    Examples
+    --------
+    >>> import collections
+    >>> from pandas.api.types import is_hashable
+    >>> a = ([],)
+    >>> isinstance(a, collections.abc.Hashable)
+    True
+    >>> is_hashable(a)
+    False
+    """
+    cdef:
+        bint is_none
+        bint is_long
+        bint is_float
+        bint is_unicode
+        bint is_tuple
+        bint is_frozen_set
+        bint is_dict
+        bint is_list
+        bint is_any_set
+
+    # Perform all checks in order to be nice to the branch predictor
+    is_none = obj is None
+    is_long = PyLong_CheckExact(obj)
+    is_float = PyFloat_CheckExact(obj)
+    is_unicode = PyUnicode_CheckExact(obj)
+    is_tuple = PyTuple_CheckExact(obj)
+    is_frozen_set = PyFrozenSet_CheckExact(obj)
+    is_dict = PyDict_CheckExact(obj)
+    is_list = PyList_CheckExact(obj)
+    is_any_set = PyAnySet_CheckExact(obj)
+
+    if is_none or is_long or is_float or is_unicode:
+        return True
+
+    # tuple and frozenset is hashable if and only if all elements are hashable
+    if is_tuple:
+        for o in <tuple>obj:
```

Comment:
Does the cast actually matter? I again would think Cython would just always defer to `PySequence_GetItem` for iteration. 

If that's true, I suppose there could be some difference between `PySequence_GetItem` and a more exact `PyTuple_GETITEM` call that could be dispatched to, with the latter not having to deal with reference counting. However, the point of writing in Cython  is to happily trade some performance compared to the raw C code for some higher level abstractions. 

Where that line really matters is up for debate, but I generally worry about trying to be too exact in this function. I'm also a little unclear why the non-hashable items can't short-circuit to quickly throw an error during a `PyObject_Hash` call. I would think the Python built-in tuple and frozen set should be able to do this faster than us?

Reply:
Thanks!

I've looked into the generated C-code for this version with the casts and a version without the cast. Assuming that the C-compiler is able to hoist some conditions out of the generated loops, the generated assembly for the two versions should be exactly the same. I went too far down the path of micro optimisations.

As to whether the built-in `tuple` and `frozenset` can do this faster than us I've looked into the implementations in CPython. `tuple` essentially call the hash function on all the elements in a loop and do some bit-twiddling to combine the hashes into one hash. The C-code of that loop is very similar to that generated by cython but we save calls to `PyObject_Hash(...)` on the elements which on the micro level should give a nice speed-up. For `frozenset` I've discover that it always is hashable, which makes perfect sense as elements must hashable to be put in a `set`. This discovery allows me to make some simplifications to the code.

<details>
 <summary>Generated C-code</summary>
 <h3>With casts</h3>

![image](https://github.com/user-attachments/assets/e386196f-ce70-45e0-96ec-ff9aeb275c84)
![image](https://github.com/user-attachments/assets/e5343658-64f3-4836-a437-ed0d5a9aabca)
 <h2>Without casts</h2>
<img width="1281" alt="image" src="https://github.com/user-attachments/assets/6c56f5a9-d904-4af7-91b4-815b273b339d">
</details> 

<details>
<summary>CPython implementation</summary>

<h3>tuple</h3>

https://github.com/python/cpython/blob/0cba289870d5cd41f24b2f63b9480e4593aa2330/Objects/tupleobject.c#L318-L342

Note `PyTuple_GET_ITEM(...)` is essentially just indexing into the underlying array
https://github.com/python/cpython/blob/0cba289870d5cd41f24b2f63b9480e4593aa2330/Include/cpython/tupleobject.h#L27

</details>